### ISSUE TITLE
allow chef client to log to syslog instead of just files

### ISSUE THREAD

**Issue by kamaradclimber** [chef/chef#1609]
Previous workaround do not work anymore due to MonoLogger class addition.

**Comment by jalessio** [comment]
I'm interested in looking into this one:
* Are there any existing proposed solutions to this?
* What was the "previous workaround" that @kamaradclimber mentions in the ticket? @stevendanna indicated that perhaps it's this? https://github.com/opscode-cookbooks/chef-client#configuration-includes
* Based on the implementation above, would it be acceptable to introduce the `syslog-logger` as a new external dependency? My guess is that it would be preferable to instead use [`Syslog::Logger`](http://ruby-doc.org/stdlib-2.0/libdoc/syslog/rdoc/Syslog/Logger.html) from the Ruby stdlib directly.
* How would you specify syslog as the desired logger? Would you set `Chef::Config[:log_location]` to something like `:syslog`? Or introduce a new config option in `client.rb`?
* A *very* quick look through the code leads me to believe that the change might actually need to go in [`chef/mixlib-log/lib/mixlib/log.rb`](https://github.com/chef/mixlib-log/blob/50ec55964ce19d3a8a14050be9a23c4b8990e2f0/lib/mixlib/log.rb#L150)? The thought being that we'd catch the special `:syslog` case and instead of returning an instance of `Logger` we'd return an instance of [`Syslog::Logger`](http://ruby-doc.org/stdlib-2.0/libdoc/syslog/rdoc/Syslog/Logger.html). Or would we want to catch the `:syslog` case before it ever hit `mixlib-log`? (*or something else completely?*)

Also, there was mention in #chef that @scotthain might have more information on this so tagging Scott here.

**Comment by kamaradclimber** [comment]
@jalessio we are currently syslogging chef logs.
The setup is done in a wrapper of chef-client cookbook which:
- chef_gem 'syslog-logger'
- create a syslog.rb file in /etc/chef/client.d directory
- reload chef-client service

The syslog file is a hack that parse chef logger output:
```
require 'rubygems'
require 'syslog'

class ChefLogParser

  def initialize
    @regexp = Regexp.new('\[(?<time>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\+\d{2}:\d{2})\] (?<crit>\w+): (?<msg>.*)')
  end

  def parse(msg)
    match = @regexp.match(msg)
    if match
      criticity = case match['crit']
      when 'DEBUG'
        Syslog::LOG_DEBUG
      when 'INFO'
        Syslog::LOG_INFO
      when 'WARN'
        Syslog::LOG_WARNING
      when 'ERROR'
        Syslog::LOG_ERR
      when 'FATAL'
        Syslog::LOG_CRIT
      else
        Syslog::LOG_CRIT
      end
      [criticity, match['msg']]
    else
      [Syslog::LOG_CRIT, msg]
    end
  end
end

class SyslogDevice
  attr_accessor :sync

  def initialize
    @parser = ChefLogParser.new
    if not self.class.const_defined?(:SYSLOG)
      self.class.const_set(:SYSLOG, Syslog.open('chef-client', nil, Syslog::LOG_DAEMON))
    end
  end

  def write(msg)
    parsed = @parser.parse(msg)
    (self.class.const_get :SYSLOG).log(parsed[0], parsed[1])
  end

  def close
    # syslog will be closed at most once since there is only one ref to it
    (self.class.const_get :SYSLOG).close
  end
end

log_location SyslogDevice.new
```

Not really elegant, I wish the logging mechanism in chef were more pluggable.

**Comment by thommay** [comment]
This is fixed in 12.4.0, due to release next week!
